File: src/cpp/readbytes2.cpp
#include "stopwatch.h"
#include <iostream>
#include <sys/mman.h>
#include <sys/stat.h>
#include <stdio.h>
#include <fcntl.h>

class MemoryMappedFileReader {
private:
  int fFile;
  unsigned char* fBuffer;
  size_t fLength;
  size_t fPos;

public:
  MemoryMappedFileReader(std::string s) : fFile(open(s.c_str(), O_RDONLY)) {
    assert(fFile != -1);
    struct stat stats;
    fstat(fFile, &stats);
    fLength = stats.st_size;
    fBuffer = static_cast<unsigned char*>(mmap(0, fLength, PROT_READ, MAP_PRIVATE, (int)fFile, 0));
    fPos = 0;
  }

  ~MemoryMappedFileReader() {
    munmap(fBuffer, fLength);
    close(fFile);
  }

  int read() {
    if (fPos >= fLength) {
      return -1;
    }
    return fBuffer[fPos++];
  }
};

size_t readBytes() {
  size_t res = 0;
#if VARIANT == 2
  size_t sum = 0;
#endif
  for (int i=0; i<10; i++) {
    MemoryMappedFileReader r("/tmp/shop_with_ids.pb");
    int read = r.read();
    while (read != -1) {
      ++res;
#if VARIANT == 2
      sum += read;
#endif
      read = r.read();
    }
  }
#if VARIANT == 2
  return res + sum;
#else
  return res;
#endif
}

int main(int argc, char** args) {
  StopWatch sw;
  sw.start();
  size_t count = readBytes();
  sw.stop();
  std::cout << "<tr><td>" << V << "-2(" << (VARIANT == 1 ? "count":"sum") << ")</td><td>" << count << "</td><td>" << sw.delta() << "</td><td>straight forward implementation using mmap.</td></tr>" << std::endl;
  return 0;
}

File: src/java/ReadBytes2.java
import java.io.FileInputStream;
import java.io.InputStream;

public class ReadBytes2 {

  static class FileReader {
    private InputStream fInput;
    private byte[] fBuffer = new byte[8192];
    private int fPos = 0;
    private int fAvailable = 0;

    public FileReader(InputStream input) throws Exception {
      fInput = input;
    }

    int read(byte[] target) throws Exception {
      if (fPos == fAvailable) {
        fAvailable = fillBuffer();
        if (-1 == fAvailable) {
          return 0;
        }
      }
      target[0] = fBuffer[fPos++];
      return 1;
    }

    private int fillBuffer() throws Exception {
      fPos = 0;
      return fInput.read(fBuffer);
    }
  }
  
  private static int readBytes() throws Exception {
    int count = 0;
    for (int i = 0; i < 10; i++) {
      FileReader in = new FileReader(new FileInputStream("/tmp/shop_with_ids.pb"));
      byte[] buffer = new byte[1];
      int c = in.read(buffer);
      while (c == 1) {
        ++count;
        c = in.read(buffer);
      }
    }
    return count;
  }

  public static void main(String[] args) throws Exception {
    long start = System.currentTimeMillis();
    int count = readBytes();
    long end = System.currentTimeMillis();
    System.out.println("Time for " + count + " bytes: " + (end - start));
  }

}
